compiler/rustc_abi/src/lib.rs

```diff
@@ -349,13 +377,63 @@ impl TargetDataLayout {
                 ["f32", a @ ..] => dl.f32_align = parse_align(a, "f32")?,
                 ["f64", a @ ..] => dl.f64_align = parse_align(a, "f64")?,
                 ["f128", a @ ..] => dl.f128_align = parse_align(a, "f128")?,
-                // FIXME(erikdesjardins): we should be parsing nonzero address spaces
-                // this will require replacing TargetDataLayout::{pointer_size,pointer_align}
-                // with e.g. `fn pointer_size_in(AddressSpace)`
-                [p @ "p", s, a @ ..] | [p @ "p0", s, a @ ..] => {
-                    dl.pointer_size = parse_size(s, p)?;
-                    dl.pointer_align = parse_align(a, p)?;
+                [p, s, a @ ..] if p.starts_with("p") => {
+                    // Some targets, such as CHERI, use the 'f' suffix in the p- spec to signal that
+                    // they use 'fat' pointers. The resulting prefix may look like `pf<addr_space>`.
+                    let p = p.trim_start_matches(char::is_alphabetic);
+
+                    let addr_space = if !p.is_empty() {
+                        parse_address_space(p, "p")?
+                    } else {
+                        AddressSpace::ZERO
+                    };
+
+                    let pointer_size = parse_size(s, p)?;
+                    let info = AddressSpaceInfo {
+                        pointer_index: pointer_size,
+                        pointer_size,
+                        pointer_align: parse_align(a, p)?,
+                    };
+                    if addr_space == default_address_space {
+                        dl.default_address_space_info = info;
+                    } else {
+                        match dl.address_space_info.iter_mut().find(|(a, _)| *a == addr_space) {
+                            Some(e) => e.1 = info,
+                            None => {
+                                dl.address_space_info.push((addr_space, info));
+                            }
+                        }
+                    }
+                }
+                [p, s, _pr, i, a @ ..] if p.starts_with("p") => {
```

Comment:
```suggestion
                [p, s, _pr, a, idc @ ..] if p.starts_with("p") => {
```
Index size is the last component?

Reply:
Thanks, I missed that. Do you think it should it be, then,
```
[p, s, a , _pr, i @ ..] if p.starts_with("p") => {
```
instead? 

For reference, the relevant bit from LLVM's datalayout string spec: 
```
p[n]:<size>:<abi>[:<pref>[:<idx>]]
```

I can also add the `pref` field to `AddressSpaceInfo`, which perhaps would make even more sense.

Reply:
commit https://github.com/rust-lang/rust/pull/143182/commits/ccc2083c1a7f6b7e466bb76ec2d195ad7002c015 adds the parsing of `:<pref>` with functions to retrieve that value: I set it to default to `<abi>` when not specified. 

I also had to slightly change the utility `parse_align` function, which took as parameter a `&[&str]`, whose semantics were actually to ignore the potential `:<pref>:<more...>` following `<abi>`.